src/server.ts

```diff
@@ -212,234 +223,256 @@ export class SubscriptionServer {
+      }
+      operationContext.executionIterator = undefined;
       if (this.onOperationComplete) {
         this.onOperationComplete(connectionContext.socket, opId);
       }
     }
   }
 
   private onClose(connectionContext: ConnectionContext) {
-    Object.keys(connectionContext.operations).forEach((opId) => {
+    connectionContext.operations.forEach((_, opId) => {
       this.unsubscribe(connectionContext, opId);
     });
   }
 
+  private async processStart(
+    connectionContext: ConnectionContext,
+    operationContext: OperationContext,
+    opId: string,
+    message: OperationMessage,
+  ) {
+    try {
+      operationContext.processingStart = true;
+      const { payload } = message;
+      if (!payload) {
+        throw new Error('Payload missing');
+      }
+      const { query } = payload;
+      if (!query) {
+        throw new Error('query missing');
+      }
+      const initResult = await connectionContext.initPromise;
+      const baseParams: ExecutionParams = {
+        query,
+        variables: payload.variables || {},
+        operationName: payload.operationName,
+        context: isObject(initResult) ? Object.assign(Object.create(Object.getPrototypeOf(initResult)), initResult) : {},
+        schema: this.schema,
+      };
+      let promisedParams = Promise.resolve(baseParams);
+
+      if (this.onOperation) {
+        promisedParams = Promise.resolve(this.onOperation(message, baseParams, connectionContext.socket));
+      }
+      const params = await promisedParams;
+      if (!isObject(params)) {
+        throw new Error('Invalid params returned from onOperation! return values must be an object!');
+      }
+      if (!params.schema) {
+        const error =
+          'Missing schema information. The GraphQL schema should be provided either statically in' +
+          ' the `SubscriptionServer` constructor or as a property on the object returned from onOperation!';
+        throw new Error(error);
+      }
+
+      const document = isString(params.query) ? parse(params.query) : params.query;
+      let executionPromise: Promise<AsyncIterator<ExecutionResult> | ExecutionResult>;
+      const validationErrors = validate(params.schema, document, this.specifiedRules);
+      if (validationErrors.length > 0) {
+        executionPromise = Promise.resolve({ errors: validationErrors });
+      } else {
+        let executor: SubscribeFunction | ExecuteFunction = this.execute;
+        if (this.subscribe && isASubscriptionOperation(document, params.operationName)) {
+          executor = this.subscribe;
+        }
+        executionPromise = Promise.resolve(
+          executor(params.schema, document, this.rootValue, params.context, params.variables, params.operationName),
+        );
+      }
+      const executionResult = await executionPromise;
+      const executionIterable = isAsyncIterable(executionResult) ? executionResult : createAsyncIterator([executionResult]);
+      if (operationContext.executionIterator) {
```

Comment:
Can you add a comment here that explains why we do this check _after_ running the execution promise?